### PR TITLE
Compile against supported openssl version

### PR DESCRIPTION
**Pull request by cryptopatrick**
## Purpose
The crate [`openssl`](https://docs.rs/openssl/0.10.32/openssl/index.html) has a struct called [`SslOptions`](https://docs.rs/openssl/0.10.32/openssl/ssl/struct.SslOptions.html) that provides __bit flags__.
These flags can be used to controll the behavior of an [`SslContext`](https://docs.rs/openssl/0.10.32/openssl/ssl/struct.SslContext.html).

> Bit Flags (also: Bit Fields) can be used to store more than one  
boolean value in one byte or a whole set of bytes.

The use of one particular bit flag is currently causing a compile   
error on machines running Ubuntu 18.04 LTS, that tries to build  
FerrumFIX without having OpenSSL version 1.1.1 or newer installed. 

This pull request removes the build error on Ubuntu 18.04 LTS 
by adding conditional compilation to the lines of code that are causing the error.

### Problem
In file [`fefix/src/fixs/mod.rs`](https://github.com/neysofu/ferrum-fix/blob/f890896c18ec91f8d5cae8ab08bef4e098439ad7/fefix/src/fixs/mod.rs#L91) the SslOptions bit flag [`NO_TLSV1_3`](https://docs.rs/openssl/0.10.32/openssl/ssl/struct.SslOptions.html#associatedconstant.NO_TLSV1_3)  
is used to disable the use of TLSv1.3.

However, on Ubuntu 18.04 LTS, the use of bit flag `NO_TLSV1_3` 
introduces a dependency on the target machine - requiring OpenSSL 1.1.1 
or newer to be installed. If OpenSSL 1.1.1 or newer is not installed, then the 
Rust compiler will __fail__  with the  __error message__:

```bash
error[E0599]: no associated item named `NO_TLSV1_3` found for struct `openssl::ssl::SslOptions` in the current scope
   --> fefix/src/fixs/mod.rs:118:49
    |
118 |                 context.set_options(SslOptions::NO_TLSV1_3);
    |                                                 ^^^^^^^^^^
    |                                                 |
    |                                                 associated item not found in `openssl::ssl::SslOptions`
    |                                                 help: there is an associated constant with a similar name: `NO_TLSV1_1`

error: aborting due to previous error
```

## Approach
This Pull Request provides an easy solution to the described problem  
by using conditional compilation, as proposed by @neysofu .  

By adding a _configure attribute_ to the `NO_TLSV1_3` bit flag, the Rust  
compiler will ignore the bit flag unless OpenSSL 1.1.1 or newer is  
installed on the machine at compile time.

```rust
// In file: fefix/src/fixs/mod.rs
#[cfg(openssl111)]
context.set_options(SslOptions::NO_TLSV1_3);
```

The solution has been tested by the following procedure:
1. Fork the latest ferrum-fix commit on branch `develop` (commit: [dd03638](https://github.com/neysofu/ferrum-fix/commit/dd036389252ed932b3f88e59ff258541563dc80b) )
2. Git clone and run `cargo build` locally
3. Verify that `cargo build` fails due to openssl dependency error
4. Make changes in a seperate branch
5. Run `cargo build` and verify that openssl error disapears

#### Open Questions and Pre-Merge TODOs
This pull request was only tested on Ubuntu version 18.04 LTS.
Some open (and potentially high priority) questions  remain:
- [ ] Does the same `cargo build` error occur on machines running LibreSSL?
- [ ] If the problem would turn out to exists on machines running LibreSSL - does the 
solution proposed in this pull request solve the issue? 
- [ ] Is TLS 1.3 supported in LibreSSL, and if so, from which version and onward?
- [ ] Does the same problem exist for machines running Windows?
- [ ] Does support for TLS 1.3 varry between clients and servers? 

## Learning
Preparation for this pull request mostly consisted of reading about 
conditional compilation and getting acquainted with [openssl](https://docs.rs/openssl/0.10.32/openssl/index.html).

### Sources used
+   [Conditional compilation](https://doc.rust-lang.org/stable/reference/conditional-compilation.html) 
+   Crate [openssl](https://docs.rs/openssl/0.10.32/openssl/index.html)
+   Bit Flag [`NO_TLSV1_3](https://docs.rs/openssl/0.10.32/openssl/ssl/struct.SslOptions.html#associatedconstant.NO_TLSV1_3)
+   LibreSSL 3.1.1 [release notes](https://ftp.openbsd.org/pub/OpenBSD/LibreSSL/libressl-3.1.1-relnotes.txt)

### CODE REVIEW

File: fefix/src/fixs/mod.rs [hunk 1]
```diff
@@ -79,6 +79,8 @@ impl Version {
                     .unwrap();
                 context.set_options(SslOptions::NO_COMPRESSION);
                 context.set_options(SslOptions::NO_SESSION_RESUMPTION_ON_RENEGOTIATION);
+                // NO_TLSV1_3 requires openssl version 1.1.1
+                #[cfg(openssl111)]
```
Comment (neysofu) [inline fefix/src/fixs/mod.rs:92]: `openssl111` needs to be available to crate users via `Cargo.toml` or detected and set automatically during build time by `build.rs`. As of now, this attribute has no effect whatsoever.

**Comment by neysofu** [comment]
Hi @cryptopatrick, thanks for looking into this and taking the time to make this PR.

Compilation failure due to an outdated version of OpenSSL is clearly undesirable behavior, so I agree that the current situation needs fixing. Let's think about this. We need to take into considerations several factors:

1. Each revision of the FIXS standard mandates a specific range of protocol versions, e.g. TLS 1.2/1.3 in the case of FIXS v1.1-rc1. Currently, FerrumFIX only supports v1.0-draft, but that will change very soon. I want our solution to be future-proof and not specific to the current revision of the FIXS standard. <sup>[1](https://www.fixtrading.org/standards/fixs/)</sup>
2. Each protocol version in turn requires a specific minimum version of OpenSSL, e.g. TLS 1.3 requires OpenSSL 1.1.1+. <sup>[2](https://wiki.openssl.org/index.php/TLS1.3)</sup>
3. OpenSSL dependency should *not* be turned on by default, but rather be opt-in via a crate feature. The same applies for `rustls`, if and when we decide to add support.

Given these constraints, I would also like to note that modifying the behavior of the library based on the version of OpenSSL is currently installed is bad security practice. Sure, the `openssl111` approach is innocuous enough, but it's quite the slippery slope, don't you think? What happens to `openssl111` once FIXS suggests TLS 1.3+ for all users? Will we disable it for people who haven't selected `openssl111` despite obvious security concerns?

Please note that besides the feature itself we also need to make it available to users in [`fefix/Cargo.toml`](https://github.com/neysofu/ferrum-fix/blob/2c266635b930abaf754969af5ba133a4a00c6983/fefix/Cargo.toml). Marking regions of the code with `#[cfg(feature = "..."])` is not enough. See the related code review above.

### My proposal

Rather that exposing weird, low-level versioning details of OpenSSL via features like `openssl111`, let's create crate features like the following:

- `fixs10`, `fixs11` (depending on what version of FIXS users want).
- `fixs-util-openssl`, which enables OpenSSL-specific features for the selected FIXS version.

This way, we avoid dealing with OpenSSL versioning issues almost entirely and we ask users to update if and only if they wish to use the most recent features, like `fixs11` and any future versions.

# Pressing issues

- [ ] Does the same cargo build error occur on machines running LibreSSL?

Not sure, I'll have to investigate and set up some tests to run on the CI server.

- [x] If the problem would turn out to exists on machines running LibreSSL - does the
solution proposed in this pull request solve the issue?

The `openssl` crate supports LibreSSL, so things should work out fine. They might require additional tuning and fixes though, but I'm cautiously optimistic. Same answer as above!

- [ ] Is TLS 1.3 supported in LibreSSL, and if so, from which version and onward?

3.2.2, but it doesn't expose `OPENSSL_NO_TLS1_3`. I'll have to think about what to do about this.

- [x] Does the same problem exist for machines running Windows?

Assuming OpenSSL 1.1.1+ is not installed, yes. The solution is the same, too.

- [x] Does support for TLS 1.3 varry between clients and servers?

OpenSSL supports both, which is what I'd like to solve with this PR. I'll create an issue to discuss LibreSSL and rustls in the future.

**Comment by neysofu** [comment]
I will expose both a `fixs` and a `fixs-util-openssl` feature, each disabled by default. `fixs` enables the `fixs` module, i.e. basic functionality like a list of suggested ciphersuites. `fixs-util-openssl`, on the other hand, adds the dependency to OpenSSL and **mandates** a recent version. I've thought long and hard about this and anything else would pose a security risk to users. The February 2021 revision of FIXS requires using the last two available TLS versions, i.e. TLS 1.2 & 1.3. I don't want users to use FIXS within FerrumFIX and thinking they're safe unless they're using the protocol versions officially supported by FIXS.

Screenshot from 2021-05-07 15-23-01

**Comment by neysofu** [comment]
With `615aec3`, OpenSSL becomes opt-in for users and opt-out for contributors (i.e. people like you @cryptopatrick and me), so that OpenSSL and TLS 1.3 -related tests are enabled by default. This is the most-security conscious and user-friendly approach imho.

What do you think? @cryptopatrick

**Comment by cryptopatrick** [comment]
@neysofu 
I really like the opt-in/opt-out decision. It makes a lot of sense to me and strikes a nice balance between 1) building for the future, while still 2) catering to the reality of users being slow to upgrade their systems.

I think that it might be quite helpful to some users of Ferrum if we made an effort to clearly document/explain the FIXS / TLS1.2/TLS1.3  / OpenSSL relationship and the decisions/trade-offs that users of the library need to be aware of regarding this topic.

**Comment by neysofu** [comment]
Awesome, I'll keep this PR open and commit some documentation for the `fixs` module as soon as tomorrow.

**Pull request merged**